lib/bitcoin/BitcoinProcessor.ts

```diff
@@ -704,4 +717,51 @@ export default class BitcoinProcessor {
     // non sidetree transaction
     return sidetreeTxToAdd;
   }
+
+  /**
+   * Return transactions grouped by block height. EX: [[transactions in block1], [transactions in block2], [transactions in block3]]
+   * @param since Transaction number to query since
+   * @param numOfBlocks number of blocks worth of transactions to get at least
+   */
+  private async getAtLeastPageSizeBlocksWorthOfTransactionsSinceTransactionNumber (since: number | undefined): Promise<TransactionModel[][]> {
+    let beginTransactionNumber = since === undefined ? 0 : TransactionNumber.getBlockNumber(since);
+    let endTransactionNumber = beginTransactionNumber + BitcoinProcessor.pageSizeInBlocks;
+
+    const groupedTransactionsToReturn: TransactionModel[][] = [];
+
+    // while need more blocks and have not reached the processed block
+    while (groupedTransactionsToReturn.length < BitcoinProcessor.pageSizeInBlocks && beginTransactionNumber <= this.lastProcessedBlock!.height) {
+      let transactions: TransactionModel[] = await this.transactionStore.getTransactionsByTransactionTime(beginTransactionNumber, endTransactionNumber);
+
+      transactions = transactions.filter((transaction) => {
+        // filter anything greater than the last processed block because they are not complete
+        return transaction.transactionTime <= this.lastProcessedBlock!.height &&
+          // if there is a since, filter transactions that are less than or equal to since (the first block will have undesired transactions)
+          (since === undefined || transaction.transactionNumber > since);
+      });
+
+      const transactionsGroupedByTransactionTime: TransactionModel[][] = BitcoinProcessor.groupTransactionsByTransactionTime(transactions);
```

Comment:
Clarifying question: since we are flattening all the transactions eventually, why do we need to group? Maybe we can do a sort-by-transaction-time at the end to achieve the same effect?

Or is it because we are using the 'grouping' to figure out the number-of-blocks that we have gotten? In that case can we use a Set/HashSet to figure out the distinct blocks and keep a running count or something?

I have a feeling that having a flat array will be more maintainable and order wouldn't change.

Reply:
I am using grouping to figure out the number of blocks that we have gotten. A flat array and keeping a count for number of blocks works too. I just thought this makes the most sense conceptually, but I'm open to change it.